### ISSUE TITLE
Need some improvement in keep-last algorithm

### ISSUE THREAD

**Issue by 0bi-w6n-K3nobi** [allinurl/goaccess#1927]
Hello, @allinurl.

I have a medium `LOG` to process. About _1.8 million_ of records more accurately.
However, it has percularity time distribution, due to my carelessness.
And, It should only have 1 day of interval. I.E `15/Sep`.

It has the dates, in chronological order:
* First from `12/Aug` until `15/Sep`.
* Soon after, it returns to `29/Aug` until `15/Sep`.
* And in third it returns to `09/Sep` until `15/Sep`.

Truly speaking -- it is the union of other logs and therefore have this particular feature.
And only _12,000_ of records do _not belong_ to `15/Sep`.

Running `GoAccess` with the option` keep-last == 1`, the program slows down  processing, 
from `20,000/sec` to `10/sec`, when it reaches `15/Sep` at first time .

Debbuging... I found that long time is spended with `clean_old_data_by_date` routine.

To compare, the processing time is:
* `09":15'` -- with the option `keep-last == 1`;
* `02":14'` -- without restriction, `keep-last == 0`;
* `02":08'` -- with `awk` filter, choosing only `15/Sep` [ restrictions do not matter here ].

**Comment by allinurl** [comment]
Thanks for reporting this! 

I was looking at this and I'm able to reproduce the issue on my end. Like you said, the issue appears to be on that function. A quick first impression it appears to be when attempting to the delete those entries from the hash table. Though, I need to investigate more on this. Feel free to post any finding from your end :)

**Comment by allinurl** [comment]
By the way, it looks like the log size has something to do cause I can't reproduce it on smaller logs.

**Comment by 0bi-w6n-K3nobi** [comment]
@allinurl , the correct term would not be small `LOG`, but few records.
The `hash table` must be quite large to have this problem.
That's what I can see.

**Comment by 0bi-w6n-K3nobi** [comment]
@allinurl , good morning.

I was playing with the same `LOG`, thinking about the `HASH` tables.

I made a transformation of the records with dates before `15/Sep`.
I.E. An elaborate filter made with `awk` to rescale to`14/Sep` in the range of `00:00:00` until `23:59:59`.

In processing, I got the following times:

* `03":08'` -- without restriction, `keep-last == 0`;
* `14":14'` -- with the option `keep-last == 1`;

I checked the # of processed records. All are OK at `15/Sep` and the rest at `14/Sep`.

Correct me if I'm wrong about what I understand about the program's behavior:
As the processing of the records advances [ chronologically ] in `15/Sep`,
it needs to _remove_ from the `HASH` table what doesn´t correspond to the _1 day interval_ [ this is our case ].

For me, it's ok. And perhaps isn't a BUG, but just how to `HASH` table was implemented.
It's all of it **make sense for real-time**. 

**Sugestion:**
But, I think that you can be a little optimization, _removing_ [ or moving ] in case normal `LOG` processing
and apply it at _ending_.
And it can be enabled/disabled by `CONF` flag.

Right; `GoAccess` will be consume more memory [ be hungry pephaps ], paying a price for faster processing.

**PS:**
And so, at ending, add a message indicating the action, Ex. :
* `Applyng keep-last constrition/filter`;
* `Processing  IP resolver`; 
* `etc`

for final user understand what's happing.

**Comment by 0bi-w6n-K3nobi** [comment]
I'm here again.

I should like shared a small patch with 2 ideas, just implementing that was commented above.

Firstly, to make me understand, I will appoint the `LOOP` in `parse.c` around line `2790` of the `BIG-LOOP` [ `BL` ].
 
The main idea is disable the call `clean_old_data_by_date` route in `BL`, if `GoAccess` not running in real-time [ `RT` ].
And after BL than running `clean_old_data_by_date` [ `CLOD` ] route.
I.E. If  `GoAcess` is running in `RT` the old behavior will be preserved.

I implemented this here. The patch that is **attached**.
And for the `LOG` mentioned here, in the ending it only took **4s** to process the `CLOD` routine.
Great!

The second idea is optimization into `BL`, same running in `RT`.
Think about, I applied a litle trick, setting `conf.keep_last == 0` before `BL`.
Remember that `BL` will be running both case, `RT` or not, and that call `CLOD` every time it is needed.
It is not an elegant way, but it works.
Disregard if you don't find it necessary.
But, take a look at the idea below a global flag.

**sugestion:**
But, for optimization I think than into `BL` several small pieces of software can be disable.
For Ex. -- In #1913 for patch that I created: 
If not in `RT` mode into `BL`, _do not make sense_ calculate individual line size. 
Just use `size_file` at ending, like that _old behavior_.

Honestly, I am in favor of creating a global flag like `GLog.intoBL`.
Done this way, all the optimizations we need can be applied by testing this flag.
For Ex.:
`if (!glog->intoBL && conf.keep_last > 0)`
` ` ` ` `clean_old_data_by_date ();`

**PS**: I did not implement this here.

[keep-last-patch.txt](https://github.com/allinurl/goaccess/files/5319706/keep-last-patch.txt)

**Comment by 0bi-w6n-K3nobi** [comment]
@allinurl, good aftermoon.
I apologize by last comment. It was so complex and a little confused.
But, not everything is lost. You can make use the patch for temporary workaround.
Forget about optimization around `BL`, but you can make use of optimization for non-`RT` mode,
in the `CLOD` routine and it calls.

After analyze my `LOG` more carefully, I found the problem.
I'll try to explain it through the analysis of the records.
See below the detailed distribution of records by date:

```
       #     Total     Date
....
       51     10369 11/Sep/2020
       56     10425 12/Sep/2020
       79     10504 13/Sep/2020
       41     10545 14/Sep/2020
       14     10559 15/Sep/2020
     1285     11844 13/Sep/2020
     1427     13271 14/Sep/2020
   621163    634434 15/Sep/2020
      214    634648 09/Sep/2020
      155    634803 10/Sep/2020
       52    634855 11/Sep/2020
      130    634985 12/Sep/2020
      523    635508 13/Sep/2020
      593    636101 14/Sep/2020
  1217298   1853399 15/Sep/2020
```
Carefully inspecting the `CLOD` routine, we can see that it in turn calls 
the `invalidate_date` routine and the `rebuild_rawdata_cache` routine.
The `invalidate_date` routine will be remove all data [`HASH`] of all `modules`/`panels`
and `rebuild_rawdata_cache` routine will be rebuild them.

All of this is ok, is very fast, until 634434 record, and the system starts to slow down.
Why?
In this point, the `HASH` has more than 621163 records in `15/Sep`. And therefore slow to rebuild it.
And for **each record**, with **date less that** `15/Sep`, will be force the `GoAccess` to same behavior.

**solution:**
Simply avoid inserting a record that will soon be invalidated because of its date.
In a practical way, apply a similar algorithm to the one below:
```
function can_insert_to_hash() {

  if (!conf.keep-last)                  /* Ok, No restrictions. */
    return 0;

  if (record_date in hash_dates)        /* Sure! Just insert. */
    return 1;

  if (conf.keep-last > len(hash_date))  /* Ok, There is a place for that. */
    return 2;

  if (record_date > min(hash_dates) {   /* Ok, Minor date was removed */
    invalidate_date(min(hash_dates));
    rebuild_rawdata_cache();
    return 3;
  }

  return -1;                            /* Oops, Not everything is going well here. */
}
```
That's all for now folks.
Feel free if you have any questions.

**Comment by allinurl** [comment]
@0bi-w6n-K3nobi Thanks for these findings. So just I understand better, you are saying that it always occur after inserting 634434 records?

**Comment by 0bi-w6n-K3nobi** [comment]
Yes, that's it. **In case of my `LOG`.**
Because before that, [record 634434] although the same process occurs [rebuilding `HASH`], 
it is very fast and cannot be noticed.

If you want to reproduce it is simple:
a) Set `conf.keep-last == 1`;
b) Loads the system with _lot of records_ on the _same date_;
c) Right after, try to load with records but with _date_ with values that are _before the 1st load_;

The system will be **slow down** in step c), for **each record**.

**Comment by 0bi-w6n-K3nobi** [comment]
Oh man! Sorry for my bad english. I will try be more clear.

Well, return to my `LOG` like a example. Assuming I have `keep-last == 1`.
Following I will show each group of records [by same date] and `GoAcces` behavior will be have when processing them:

```
       #     Sum     Date
....
       51     10369 11/Sep/2020  --  Just processing them.
       56     10425 12/Sep/2020  --  [ oneIAR ] In 1st record, call invalidate_date(11/Sep)
                                       and rebuild_rawdata_cache,
                                       after just he processing of their remains.
       79     10504 13/Sep/2020  --  oneIAR (12/Sep)
       41     10545 14/Sep/2020  --  oneIAR (13/Sep)
       14     10559 15/Sep/2020  --  oneIAR (14/Sep)
     1285     11844 13/Sep/2020  --  [ oopsIAR ] Oops! For each record, call invalidate_date(13/Sep) 
                                       and rebuild_rawdata_cache. Not problem here, it will be so fast.
     1427     13271 14/Sep/2020  --  oopsIAR(14/Sep)
   621163    634434 15/Sep/2020  --  Just processing them. [ no oneIAR here, because already had deleted ] 
      214    634648 09/Sep/2020  --  oopsIAR(09/Sep). So Bad, it wil be so slow. [ rebuild_rawdata_cache ]
      155    634803 10/Sep/2020  --  oopsIAR(10/Sep). So Bad, it wil be so slow.
       52    634855 11/Sep/2020  --  oopsIAR(11/Sep). So Bad, it wil be so slow.
      130    634985 12/Sep/2020  --  oopsIAR(12/Sep). So Bad, it wil be so slow.
      523    635508 13/Sep/2020  --  oopsIAR(13/Sep). So Bad, it wil be so slow.
      593    636101 14/Sep/2020  --  oopsIAR(14/Sep). So Bad, it wil be so slow.
  1217298   1853399 15/Sep/2020  --  Just processing them. [ no oneIAR here, because already had deleted ] 
```
That's all for now folks. I hope had be more clear.

**Comment by allinurl** [comment]
Thanks for posting that. 

I'm looking at this and from my understanding, the algorithm you posted above is actually very similar what it does now, did I get it right? Although, your case might be isolated as the dataset is not ordered which could affect how it performs.

**Comment by 0bi-w6n-K3nobi** [comment]
Oh... really? I didn't look for in code for some criteria for insert new record.
Sorry for that, if you already use some algorithm like in my sugestion of **solution**.

Yeah... My case is not typical, however can be occurs when `LOG` is concatenation
of small ones  than not was rotated by date but by size.
Thanks.

**Comment by allinurl** [comment]
I pushed a [**fix**](https://github.com/allinurl/goaccess/commit/0bd93e0ac796c5641a37990b3601f3311a37d7b1) that should address this issue. Could you please rebase and let me know if that solves the issue on your end. Thanks again for the details and suggestions!

**Comment by 0bi-w6n-K3nobi** [comment]
Yeah. Great man!
But I would still propose the following modification:

`clean_old_data_by_date (uint32_t numdate) {`
` ... ...`
` ` `/* if currently parsed date is in the set of its, keep inserting one */`
` ` `for (idx = 0; idx < len; idx++)`
` ` ` ` ` if (dates[idx] == numdate)`
` ` ` ` ` ` ` ` `return 1;`

This new conditional will **avoid deleting** **`date[0]`**, if `numdate` is into the current date set.
The current version will exclude `date[0]` if it is not the last date, and that should be preserved.
Thus, you will lost a date that may be important.

See you soon.

**Comment by allinurl** [comment]
@0bi-w6n-K3nobi Thanks for pointing that out. I've pushed a [commit](https://github.com/allinurl/goaccess/commit/556b608311762939c1d8f9cd4578f2158e799aa8) with it.

Let me know how it goes on your end. I'll be pushing out v1.4.1 in the upcoming days :)

**Comment by 0bi-w6n-K3nobi** [comment]
Great... Again, being extremely detailed ... you know :)

`uint32_t idx;`

Is perfect. I will test with the LOG of this issue.

**Comment by allinurl** [comment]
Great catch! I think due to the count reaching 0, using `uint32_t idx;` would throwing a warning. I just updated the loop, should work as expected now.

**Comment by 0bi-w6n-K3nobi** [comment]
Warning? Why? Zero is part of `Usigned Int` too.  ;)

**Comment by 0bi-w6n-K3nobi** [comment]
Gosh! 
Do you talk about `len - 1` into `for`?... **Missed** test `len > 0`.

Ooh! 
No problem because you test below, and return: `ht_get_size_dates () < conf.keep_last`.
And call `clean_old_data_by_date` only `conf.keep_last > 0`.
Then, `len`  will be **always** **`> 0`**.

**Comment by 0bi-w6n-K3nobi** [comment]
OK. It will be some difficult to initial programmer understand, but ok the code:

` ` `for (idx = len ; idx-- > 0 ; ) {`

But, again I talk that you already tested `len > 0` into code above.
Things and Thinks of Language C.

**Comment by allinurl** [comment]
I see what you mean. Even though is already tested, the warning: `comparison of unsigned expression in ‘>= 0’ is always true` would have persisted if I kept the loop as: 

    for (idx = len -1; idx >= 0; idx--)

I agree, it's a bit more difficult to understand, I think a simple`while (idx--) {}` would have done it.

**Comment by 0bi-w6n-K3nobi** [comment]
Huuum... I hadn't noticed the `unsigned` prank.
Things really get much more difficult. 
A good compiler would show this warning.

May be it more easy:

`dates = get_sorted_dates (&len);`

`/* If currently parsed date is in the set of dates, keep inserting it.`
` * We count down since more likely the currently parsed date is at the last pos */`
`for (idx = len; idx > 0; --idx) {`
` ` `if (dates[idx - 1] == numdate) {`
` ` ` ` `free (dates);`
` ` ` ` `return 1;`
` ` `}`
`}`

It doesn't make any difference to the compiler if you use `idx` or `idx - 1` as an index.
The optimizations will ensure the best performance.
You can take a look at the generated Assembly and compare the two versions.

I believe that this way the code for laypeople is more understandable.
There will be no loss of efficiency due to the compiler.
Everyone who is used to programming in Language C knows that every vector is zero-based.

Things and Thinks of Language C.

**Comment by 0bi-w6n-K3nobi** [comment]
Yeah. Great!

I applied all patches, I.E. the code is last version, with patch ae98eb9.
Then I runned with `--keep-last=1`. Firstly with LOG only with `15/Sep`, and after the full LOG with all dates.
See yourself the result:

![image](https://user-images.githubusercontent.com/71027865/97061842-b6a84900-156e-11eb-82f4-4fbce6bba3ff.png)

![image](https://user-images.githubusercontent.com/71027865/97061863-d2135400-156e-11eb-879b-afdb4d22e23c.png)

Great! Only about 4 seconds for filtering invalid dates.
